File: core/src/main/java/dev/bee/kanjianki/core/DictionaryLookup.java
package dev.bee.kanjianki.core;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

public abstract class DictionaryLookup {
    private static final Pattern MULTI_WHITESPACE = Pattern.compile("\\s+");

    public static final String SOURCE_KANJIDIC2 = "KANJIDIC2";
    public static final String SOURCE_ANKI = "anki";

    public abstract KanjiEntry lookupKanji(String literal);

    public abstract int kanjiCount();

    public JitenKanjiRanks jitenRanks() {
        return JitenKanjiRanks.empty();
    }

    public StudyCue studyCue(
            String kanji,
            String ankiMeaning,
            String rowReading,
            String sourceExpression,
            String sourceReading
    ) {
        KanjiEntry kanjiEntry = lookupKanji(normalize(kanji));
        String cueReading = firstNonEmpty(
                sourceReading,
                rowReading,
                kanjiEntry == null ? "" : kanjiEntry.firstReading()
        );
        String fromExpression = normalize(sourceExpression);
        if (kanjiEntry != null) {
            return new StudyCue(
                    StudyCueFormatter.displayGlosses(kanjiEntry.meanings, 2),
                    cueReading,
                    fromExpression,
                    SOURCE_KANJIDIC2
            );
        }
        return new StudyCue(
                StudyCueFormatter.cleanFallbackMeaning(ankiMeaning, "", 96),
                cueReading,
                fromExpression,
                SOURCE_ANKI
        );
    }

    public static DictionaryLookup empty() {
        return MemoryDictionaryLookup.EMPTY;
    }

    public static DictionaryLookup fromKanjiEntries(List<KanjiEntry> kanji) {
        return new MemoryDictionaryLookup(kanji);
    }

    protected static String normalize(String value) {
        return value == null ? "" : MULTI_WHITESPACE.matcher(value.trim()).replaceAll(" ");
    }

    private static String firstNonEmpty(String... values) {
        for (String value : values) {
            String normalized = normalize(value);
            if (!normalized.isEmpty()) {
                return normalized;
            }
        }
        return "";
    }

    private static final class MemoryDictionaryLookup extends DictionaryLookup {
        private static final MemoryDictionaryLookup EMPTY = new MemoryDictionaryLookup(Collections.emptyList());

        private final Map<String, KanjiEntry> kanjiByLiteral;

        private MemoryDictionaryLookup(List<KanjiEntry> kanji) {
            Map<String, KanjiEntry> byLiteral = new HashMap<>();
            for (KanjiEntry entry : kanji == null ? Collections.<KanjiEntry>emptyList() : kanji) {
                byLiteral.put(entry.literal, entry);
            }
            kanjiByLiteral = Collections.unmodifiableMap(byLiteral);
        }

        @Override
        public KanjiEntry lookupKanji(String literal) {
            return kanjiByLiteral.get(normalize(literal));
        }

        @Override
        public int kanjiCount() {
            return kanjiByLiteral.size();
        }
    }

    public static final class KanjiEntry {
        public final String literal;
        public final List<String> meanings;
        public final List<String> onReadings;
        public final List<String> kunReadings;
        public final List<String> nanoriReadings;
        public final int strokeCount;
        public final int grade;
        public final int radical;
        public final int kanjidicFrequency;
        public final Integer jitenRank;

        public KanjiEntry(KanjiEntryFields fields) {
            KanjiEntryFields safeFields = fields == null ? KanjiEntryFields.empty() : fields;
            this.literal = normalize(safeFields.literal);
            this.meanings = immutableList(safeFields.meanings);
            this.onReadings = immutableList(safeFields.onReadings);
            this.kunReadings = immutableList(safeFields.kunReadings);
            this.nanoriReadings = immutableList(safeFields.nanoriReadings);
            this.strokeCount = safeFields.strokeCount;
            this.grade = safeFields.grade;
            this.radical = safeFields.radical;
            this.kanjidicFrequency = safeFields.kanjidicFrequency;
            this.jitenRank = safeFields.jitenRank;
        }

        private static List<String> immutableList(List<String> values) {
            return Collections.unmodifiableList(new ArrayList<>(values == null ? Collections.<String>emptyList() : values));
        }

        private String firstReading() {
            if (!kunReadings.isEmpty()) {
                return kunReadings.get(0);
            }
            if (!onReadings.isEmpty()) {
                return onReadings.get(0);
            }
            return nanoriReadings.isEmpty() ? "" : nanoriReadings.get(0);
        }
    }

    public record KanjiEntryFields(
            String literal,
            List<String> meanings,
            List<String> onReadings,
            List<String> kunReadings,
            List<String> nanoriReadings,
            int strokeCount,
            int grade,
            int radical,
            int kanjidicFrequency,
            Integer jitenRank
    ) {
        private static KanjiEntryFields empty() {
            return new KanjiEntryFields(
                    "",
                    Collections.emptyList(),
                    Collections.emptyList(),
                    Collections.emptyList(),
                    Collections.emptyList(),
                    0,
                    0,
                    0,
                    0,
                    null
            );
        }
    }
}

File: dictionary-core/src/main/kotlin/dev/bee/kanjianki/core/DictionaryLookup.kt
package dev.bee.kanjianki.core

import java.util.ArrayList
import java.util.Collections
import java.util.HashMap
import java.util.Objects
import java.util.regex.Pattern

abstract class DictionaryLookup {
    abstract fun lookupKanji(literal: String?): KanjiEntry?

    abstract fun kanjiCount(): Int

    open fun jitenRanks(): JitenKanjiRanks = JitenKanjiRanks.empty()

    open fun searchKanji(query: String?, limit: Int): List<KanjiEntry> = emptyList()

    open fun eligibleKanjiCount(range: JitenRankRange): Int = 0

    open fun kanjiByJitenRank(
        range: JitenRankRange,
        offset: Int,
        limit: Int,
    ): KanjiEntryPage = KanjiEntryPage.empty()

    fun studyCue(
        kanji: String?,
        ankiMeaning: String?,
        rowReading: String?,
        sourceExpression: String?,
        sourceReading: String?,
    ): StudyCue {
        val kanjiEntry = lookupKanji(normalize(kanji))
        val cueReading = firstNonEmpty(
            sourceReading,
            rowReading,
            if (kanjiEntry == null) "" else kanjiEntry.firstReading(),
        )
        val fromExpression = normalize(sourceExpression)
        if (kanjiEntry != null) {
            return StudyCue(
                StudyCueFormatter.displayGlosses(kanjiEntry.meanings, 2),
                cueReading,
                fromExpression,
                SOURCE_KANJIDIC2,
            )
        }
        return StudyCue(
            StudyCueFormatter.cleanFallbackMeaning(ankiMeaning, "", 96),
            cueReading,
            fromExpression,
            SOURCE_ANKI,
        )
    }

    private class MemoryDictionaryLookup(kanji: List<KanjiEntry>?) : DictionaryLookup() {
        private val kanjiByLiteral: Map<String, KanjiEntry>

        init {
            val byLiteral = HashMap<String, KanjiEntry>()
            for (entry in kanji ?: emptyList()) {
                byLiteral[entry.literal] = entry
            }
            kanjiByLiteral = Collections.unmodifiableMap(byLiteral)
        }

        override fun lookupKanji(literal: String?): KanjiEntry? = kanjiByLiteral[normalize(literal)]

        override fun kanjiCount(): Int = kanjiByLiteral.size

        override fun eligibleKanjiCount(range: JitenRankRange): Int {
            if (!range.isValid()) {
                return 0
            }
            return kanjiByLiteral.values.count { entry -> range.includes(entry.jitenRank) }
        }

        override fun kanjiByJitenRank(
            range: JitenRankRange,
            offset: Int,
            limit: Int,
        ): KanjiEntryPage {
            if (!range.isValid() || offset < 0 || limit < 1) {
                return KanjiEntryPage.empty()
            }
            val eligible = kanjiByLiteral.values
                .asSequence()
                .filter { entry -> range.includes(entry.jitenRank) }
                .sortedWith(KANJI_RANK_COMPARATOR)
                .toList()
            val boundedLimit = limit.coerceAtMost(MAX_KANJI_PAGE_SIZE)
            val page = eligible.drop(offset).take(boundedLimit)
            val nextOffset = (offset + page.size).takeIf { next -> next < eligible.size }
            return KanjiEntryPage(page, eligible.size, nextOffset)
        }

        companion object {
            val EMPTY = MemoryDictionaryLookup(emptyList())
        }
    }

    data class JitenRankRange(
        val minimumRank: Int,
        val maximumRank: Int,
        val includeUnranked: Boolean = false,
    ) {
        fun isValid(): Boolean = minimumRank >= 1 && maximumRank >= minimumRank

        fun includes(rank: Int?): Boolean {
            return if (rank == null) includeUnranked else rank in minimumRank..maximumRank
        }
    }

    data class KanjiEntryPage(
        val entries: List<KanjiEntry>,
        val totalEligible: Int,
        val nextOffset: Int?,
    ) {
        companion object {
            fun empty(): KanjiEntryPage = KanjiEntryPage(emptyList(), 0, null)
        }
    }

    class KanjiEntry(fields: KanjiEntryFields?) {
        @JvmField val literal: String
        @JvmField val meanings: List<String>
        @JvmField val onReadings: List<String>
        @JvmField val kunReadings: List<String>
        @JvmField val nanoriReadings: List<String>
        @JvmField val strokeCount: Int
        @JvmField val grade: Int
        @JvmField val radical: Int
        @JvmField val kanjidicFrequency: Int
        @JvmField val jitenRank: Int?

        init {
            val safeFields = fields ?: KanjiEntryFields(
                "",
                emptyList(),
                emptyList(),
                emptyList(),
                emptyList(),
                0,
                0,
                0,
                0,
                null,
            )
            literal = normalize(safeFields.literal)
            meanings = immutableList(safeFields.meanings)
            onReadings = immutableList(safeFields.onReadings)
            kunReadings = immutableList(safeFields.kunReadings)
            nanoriReadings = immutableList(safeFields.nanoriReadings)
            strokeCount = safeFields.strokeCount
            grade = safeFields.grade
            radical = safeFields.radical
            kanjidicFrequency = safeFields.kanjidicFrequency
            jitenRank = safeFields.jitenRank
        }

        fun firstReading(): String {
            if (kunReadings.isNotEmpty()) {
                return kunReadings[0]
            }
            if (onReadings.isNotEmpty()) {
                return onReadings[0]
            }
            return if (nanoriReadings.isEmpty()) "" else nanoriReadings[0]
        }

        companion object {
            private fun immutableList(values: List<String>?): List<String> {
                return Collections.unmodifiableList(ArrayList(Objects.requireNonNullElse(values, emptyList())))
            }
        }
    }

    @JvmRecord
    data class KanjiEntryFields(
        val literal: String?,
        val meanings: List<String>?,
        val onReadings: List<String>?,
        val kunReadings: List<String>?,
        val nanoriReadings: List<String>?,
        val strokeCount: Int,
        val grade: Int,
        val radical: Int,
        val kanjidicFrequency: Int,
        val jitenRank: Int?,
    )

    companion object {
        const val MAX_KANJI_PAGE_SIZE: Int = 10_000
        private val MULTI_WHITESPACE: Pattern = Pattern.compile("\\s+")
        private val KANJI_RANK_COMPARATOR =
            compareBy<KanjiEntry> { it.jitenRank ?: Int.MAX_VALUE }
                .thenBy { it.literal }

        const val SOURCE_KANJIDIC2: String = "KANJIDIC2"
        const val SOURCE_ANKI: String = "anki"

        @JvmStatic
        fun empty(): DictionaryLookup = MemoryDictionaryLookup.EMPTY

        @JvmStatic
        fun fromKanjiEntries(kanji: List<KanjiEntry>?): DictionaryLookup = MemoryDictionaryLookup(kanji)

        @JvmStatic
        fun normalize(value: String?): String {
            return MULTI_WHITESPACE.matcher((value ?: "").trim()).replaceAll(" ")
        }

        private fun firstNonEmpty(vararg values: String?): String {
            for (value in values) {
                val normalized = normalize(value)
                if (normalized.isNotEmpty()) {
                    return normalized
                }
            }
            return ""
        }
    }
}
